src/v/storage/segment_appender.cc

```diff
@@ -445,66 +446,113 @@ void segment_appender::dispatch_background_head_write() {
 
     const size_t start_offset = ss::align_down<size_t>(
       _committed_offset, _head->alignment());
-    const size_t expected = _head->dma_size();
-    const char* src = _head->dma_ptr();
-    vassert(
-      expected <= _chunk_size,
-      "Writes can be at most a full segment. Expected {}, attempted write: {}",
-      _chunk_size,
-      expected);
+
     // accounting synchronously
+    const auto prior_co = _committed_offset;
     _committed_offset += _head->bytes_pending();
     _bytes_flush_pending -= _head->bytes_pending();
+
+    inflight_write entry{
+      .chunk = _head,
+      .chunk_begin = _head->pending_aligned_begin(),
+      .chunk_end = _head->pending_aligned_end(),
+      .file_start_offset = start_offset,
+      .committed_offset = _committed_offset};
+
     // background write
     _head->flush();
+
```

Comment:
Most of the below diff is not net new code but just re-arranged. E.g., the cleanup below moved from the bottom.